### ISSUE TITLE
Problems loading sessions

### ISSUE THREAD

**Issue by Mashin6** [igvteam/igv#829]
This is a bizarre problem, but since I upgraded to macOS Catalina (10.15.5 ), I can't load my old saved sessions. It impacts the newest, but also all previous versions of IGV that I tested. After attempting to load the session file I get error window with these messages
`/path/to/my/file.bigWig   org.broad.igv.exceptions.DataLoadException: /path/to/my/file.bigWig (Operation not permitted) `

When I dismiss the error window, IGV keeps throttling my cpu, but never succeeds loading. The error strangely disappears once I load the track file at least once manually, but after IGV restart is appears again.
I was suspecting the new security on Catalina, but even when I give IGV full disk access it still does not solve it.

**Comment by jrobinso** [comment]
Are these files under "Documents",  "Desktop", or "Downloads".

I'm sure you are correct this is Catalina security.

**Comment by jrobinso** [comment]
BTW I just got a new Catalina macbook and have tried to reproduce this
problem without success.

**Comment by jrobinso** [comment]
Reading between the lines, since nothing is explict, on some Apple message boards it might actually depend on which browser you downloading the app installer from.   I used Chrome.   What did you use?

**Comment by Mashin6** [comment]
Thanks for checking.
The track files are in Documents subfolder. There are Catalina security settings that control access to Document folders, but apparently only the App or a specific process can request that and user can't manually add. 
I also used Chrome. It's weird. Could that be a java version issue?

**Comment by jrobinso** [comment]
@Mashin6  OK that's good info,  where did you find it?  (That the App has to add).   We will see if there is a way to do that manually.

We did make some progress.   We found that we can save a session that references files in Documents, to some other folder, and load it successfully as long as we use relative paths to the files in Documents from within the session (relative to session file location).   That is now the default when saving sessions, but hasn't always been.   However, if we save sessions using absolute paths they will not load.    

BTW downloading the command-line generic version of IGV and starting it from a terminal avoids all of this.   That's not a good solution obviously.

**Comment by jrobinso** [comment]
I'm going to experiment with converting absolute paths to relative ones in code before opening a file,  so we might open as file as   '../Documents/myBam.bam.bai'  instead of  '/Users/me/Documents/myBam.bam.bai'.    Its not always possible to compute a relative path, and there are pitfalls, it would be better to just get explicit permission if we can figure out how to do that.   Since we're using Java some of this is out of our control.

**Comment by jrobinso** [comment]
Scratch the absolute vs relative path theory.  Its not repeatable.   It seems to be random in fact, sometimes I can load a session and sometimes I cannot.

The root of the problem is IGV (via Java) is not asking for permissions for these folders.  As I understand things, mostly ancedotal I have yet to find Apple documentation where this is specified in an understandable way,  apps should ask permission for access to the privileged folders when they are first acessed.   We will pursue that.   In the meantime we can't guarantee support for data files stored in privileged folders  (the 3 Ds,  Documents, Desktop, Downloads).

**Comment by jrobinso** [comment]
@Mashin6 dead end so far wrt our App package.    But here's an alternative.   Something like this might become the default.   If you want to try it and report back that would be helpful.

1. Install Java 11 if you don't have it already, I got it from here  https://adoptopenjdk.net/
2. Install our generic platform agnostic build, its labeled "IGV and igvtools to run on the command line..."
3. Unzip the archive, on my new catalina it was automatically unzipped.
4. Move the folder to the location of your choice
5. Start IGV from a terminal from the unzipped terminal with the "igv.command" script
```
./igv.command
```
Starting IGV in this way seems to work properly, at least with limited testing.

**Comment by jrobinso** [comment]
@Mashin6 actually its simpler than that, as reported in #827 you can start IGV directly from the packaged bundle is you are comfortable with the command line.   After installing Java, open a terminal and cd to /Applications or wherever you placed the IGV bundle.   Then just execute
```
IGV_2.8.6.app/Contents/MacOS/IGV
```

**Comment by Mashin6** [comment]
The last solution really works for me now. Thanks for all the debugging.  It's a bit more inconvenient, but I can make alias for /Applications/IGV_2.8.6.app/Contents/MacOS/IGV in .bash_profile and I am running terminal all the time anyways.

The info about the process access privileges I found it on this forum: https://developer.apple.com/forums/thread/125438
I feel like there must be a direct solution for Java apps. It's weird that IGV can open track when directly requested by user, but can't when batch opening from scripted path.

**Comment by jrobinso** [comment]
@Mashin6 I found I only had to run ``` /Applications/..../IGV```  once.  It prompted me to grant permission, which put IGV in the "Files and Folder" permissions list.   After that I have been able to just click the application icon as normal.    Is your experience different?  Did it prompt you for permission?

**Comment by jrobinso** [comment]
The thread posted above might have a hint of the answer.  We currently use exec in the startup script to launch Java, this could be breaking the chain of responsibility.  We'll try taking it out, but probably can't get the app notarized for testing until early next week.   Stay tuned.

**Comment by Mashin6** [comment]
I had IGV still in "full disk access", but once I removed is and ran from terminal it gave me the permission prompt. It added IGV into "files and folder permission", but when I restart app through icon it still can't access the files. 

Another bizarre thing. I ran IGV 2.4.3 (icon click) and when I tried opening a saved session, I got the prompt and the access stays even after quit and restart.
I remember the times when Apple was making fun from Vista's User Account Control. Look how turntables now.

**Comment by jrobinso** [comment]
Totally weird.   With your Mac, my colleague Helga's Mac, and my new Mac we
have 3 Catalinas.    Behavior different on all.

I'm hopeful removing the "exec" fork in our startup script might fix all
this.    I don't know why I should be hopeful.   We can't do the
notarization today to try it.

Another weird thing, if we haven't had enough yet,  IGV 2.4.3 isn't
notarized at all.   Yet it has more access than 2.8.6 which is.

**Comment by Mashin6** [comment]
Thanks for trying. I still hope that will work.

Some of that might be depending on whether it is a clean Catalina install on an upgrade. For example, Catalina now runs by default zsh, but upgraded system runs bash unless opted in zsh. Who knows what else they changed.

Not sure if this is anyhow helpful, but IGV 2.8.6 gives me in terminal few warning that I don't see with 2.4.3
```
WARNING: package com.sun.java.swing.plaf.windows not in java.desktop
WARNING: package sun.awt.windows not in java.desktop
[RuntimeUtils.java:305] [AWT-EventQueue-0]  Error loading jar: class jdk.internal.loader.ClassLoaders$AppClassLoader cannot be cast to class java.net.URLClassLoader (jdk.internal.loader.ClassLoaders$AppClassLoader and java.net.URLClassLoader are in module java.base of loader 'bootstrap')
```

**Comment by jrobinso** [comment]
@davideby Do you know the significance of the 2 warnings, and is there anything we can do about them.   More importantly, what is that runtime error (Error loading jar)?

I can imagine zsh vs bash could conceivably make a difference, because the root problem seems to be a broke responsibility chain.  To further muddy the waters I have an IT managed mac from UCSD,  my colleague Helga has an IT managed mac from the Broad Institute.   They seem to behave differently wrt these problems.

**Comment by jrobinso** [comment]
@Mashin6  We think (hope) this might be fixed with this build, if you want to give it a try.   https://data.broadinstitute.org/igv/projects/downloads/snapshot/IGV_snapshot.app.zip

**Comment by Mashin6** [comment]
Bingo! This one works at least on my computer.
Thanks a lot for the fix!

**Comment by jrobinso** [comment]
@Mashin6 Great it worked for us, but in DIFFERENT WAYS!!!   Oh well if it works in multiple ways its better than failing in multiple ways.

**Comment by davideby** [comment]
Hey @jrobinso, just to follow up your question from above, it looks like some bit of code somewhere is casting an object to `java.net.URLClassLoader` where the actual class is different.  I mean, I don't think it's anything surprising other than a standard miscast object.  Probably should just be casting to a superclass or interface above the both of those.

Unfortunately there's no line number, which is weird.  It could also be in some library code for all we can tell.  Weird that it doesn't kill IGV on load.

The two warnings can be ignored.  They're just caused by the fact that we need to enable special flags for Windows (due to JIDE requirements, IIRC) but on the Mac obviously those classes aren't present.  We see similar warnings in the other direction on Windows or Linux.

**Comment by jrobinso** [comment]
@davideby Thanks, the cast is part of an old no-longer-used plugin feature that surely wouldn't work anymore even if we wanted it to.  Will remove.

**Comment by Mashin6** [comment]
@jrobinso  This somehow crept back in. Version 2.8.7 is OK, 2.9.4 does not ask, 2.10.0 does not ask :(

**Comment by jrobinso** [comment]
@Mashin6 This thread is so long as to be basically useless WRT referring back.  What crept in?  Ask about what?  What is the specific problem, preferably with log statements if applicable.

**Comment by Mashin6** [comment]
Sorry for the brevity. What I ment to write is that when I try to open session xml file in  2.8.7 it asked me if I allow access to Documents folder. Versions  2.9.4 and 2.10.0 for some reason don't do that and when attempting to load the session I get popup with `Errors were encountered loading the session:` `org.broad.igv.exceptions.DataLoadException: somFile.bigWig (Operation not permitted)`.

And that for both with Java bundle or without Java apps.
I tried running from terminal and both of those load session fine, but still do not ask for permission.

**Comment by helgathorv** [comment]
@Mashin6  I was able to reproduce the issue on my MacBook. These Apple security issues can be difficult to track down - we'll see what we can do.

**Comment by Mashin6** [comment]
Thank you for looking into this!

**Comment by jrobinso** [comment]
@Mashin6 We are still looking for a solution but its not clear there is going to be one.  This regression occurred due to a JDK update, not a change in IGV code.   It appears to stem from loading files implicitly from a session without an explicit choice from a file dialog.   We have a similar problem with index files (e.g. bai),  the solution there is to allow the user to select the index along with the bam file from the file dialog.   Obviously we can't do that from sessions.

So at the moment we don't have a solution other than to recommend storing data in non-restricted (e.g. Documents) folders.

I'm reopening, if we find a solution we will update here.   @davideby if I've misstated anything please correct

**Comment by Mashin6** [comment]
Bummer.. Thanks for trying though.
I just mention here for those who end up here because of the same issue that another workaround is to launch IGV from Terminal e.g. `/Applications/IGV_2.10.0.app/Contents/MacOS/IGV `

**Comment by jrobinso** [comment]
This might be fixed with the latest release (2.11.0).   If not this thread is entirely too long and probably includes multiple problems, open a new issue if needed with as clear as possible instructions for steps-to-reproduce.

**Comment by malcook** [comment]
Oh goodie a new release!

Ouch:

http://software.broadinstitute.org/software/igv/2.11.x#LatestVersion   shows “Access Denied”
